mne_bids/utils.py

```diff
@@ -29,6 +30,128 @@
 
 
+class BIDSPath(dict):
+    """Create a partial/full BIDS filename from its component parts.
+
+    BIDS filename prefixes have one or more pieces of metadata in them. They
+    must follow a particular order, which is followed by this function. This
+    will generate the *prefix* for a BIDS filename that can be used with many
+    subsequent files, or you may also give a suffix that will then complete
+    the file name.
+
+    Note that all parameters are not applicable to each kind of data. For
+    example, electrode location TSV files do not need a task field.
+
+    Parameters
+    ----------
+    subject : str | None
+        The subject ID. Corresponds to "sub".
+    session : str | None
+        The session for a item. Corresponds to "ses".
+    task : str | None
+        The task for a item. Corresponds to "task".
+    acquisition: str | None
+        The acquisition parameters for the item. Corresponds to "acq".
+    run : int | None
+        The run number for this item. Corresponds to "run".
+    processing : str | None
+        The processing label for this item. Corresponds to "proc".
+    recording : str | None
+        The recording name for this item. Corresponds to "recording".
+    space : str | None
+        The coordinate space for an anatomical file. Corresponds to "space".
+    prefix : str | None
+        The prefix for the filename to be created. E.g., a path to the folder
+        in which you wish to create a file with this name.
+    suffix : str | None
+        The suffix for the filename to be created. E.g., 'audio.wav'.
+
+    Returns
+    -------
+    filename : str
+        The BIDS filename you wish to create.
+
+    Examples
+    --------
+    >>> bids_basename = make_bids_basename(subject='test', session='two', task='mytask', suffix='data.csv')  # noqa: E501
+    >>> print(bids_basename)
+    sub-test_ses-two_task-mytask_data.csv
+    >>> print(bids_basename.__repr__())
+    BIDSPath (sub-test_ses-two_task-mytask_data.csv)
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(BIDSPath, self).__init__(*args, **kwargs)
+        # run string representation to check validity of arguments
+        self.as_str()
+
+    def __setitem__(self, key, value):
+        """Set item as a dictionary, and perform validation checks."""
+        if key not in ('sub', 'ses', 'task', 'acq',
+                       'proc', 'acq', 'run', 'rec',
+                       'space', 'suffix', 'prefix'):
+            raise ValueError('Key must be one of blah, got %s' % key)
+
+        return dict.__setitem__(self, key, value)
+
+    def __str__(self):
+        """Return the string representation of the path."""
+        return self._get_name()
+
+    def __repr__(self):
+        """Representation in the style of `pathlib.Path`."""
+        return "{}({!r})".format(self.__class__.__name__, self._get_name())
+
+    def __add__(self, value):
+        """Return self+value."""
+        return self.as_str() + value
```

Comment:
if you add 2 objects of type T you should expect T + T to be still of type T. It's another reason to move away from dict. I am not sure why add is necessary here also.

Reply:
I don't get this either. Agreed about moving away from `dict` and simply setting attributes.

Reply:
Okay, so if people want to say do something like:

```
bids_fpath = bids_basename + '_ieeg.vhdr'
```

what would be the expectation (asking because this came up in the codebase)? For them to explicitly convert to a string? E.g.

```
bids_fpath = str(bids_basename) + '_ieeg.vhdr'
```

Reply:
I like option 2 better